File: backend/app/api/api_configs.py
"""API config management routes."""

import json
from fastapi import APIRouter, Query, UploadFile, File
from fastapi.responses import JSONResponse

from app.core.service_registry import registry
from app.models.response import success
from app.services.config_manager import ConfigManager

router = APIRouter(prefix="/api/api-configs", tags=["接口配置"])


def _get_mgr() -> ConfigManager:
    return registry.get("config_manager")


@router.get("")
async def list_configs(
    keyword: str | None = None,
    group: str | None = None,
    enabled: bool | None = None,
    dataSourceType: str | None = None,
):
    mgr = _get_mgr()
    configs = await mgr.list_configs(keyword=keyword, group=group, enabled=enabled, data_source_type=dataSourceType)
    return success({"items": configs, "total": len(configs)})


@router.get("/groups")
async def list_groups():
    mgr = _get_mgr()
    groups = await mgr.list_groups()
    return success(groups)


@router.get("/default-sql/{collection}")
async def get_default_sql(collection: str):
    mgr = _get_mgr()
    result = mgr.get_default_sql(collection)
    if not result:
        return success({"collection": collection, "sql": f"SELECT * FROM {collection}", "description": "默认查询"})
    return success(result)


@router.get("/export")
async def export_configs(group: str | None = None):
    mgr = _get_mgr()
    configs = await mgr.export_configs(group=group)
    return JSONResponse(content=configs, headers={
        "Content-Disposition": 'attachment; filename="api_configs.json"'
    })


@router.post("/import")
async def import_configs(body: dict):
    mgr = _get_mgr()
    configs = body.get("configs", [])
    overwrite = body.get("overwrite", False)
    result = await mgr.import_configs(configs, overwrite=overwrite)
    return success(result)


@router.post("")
async def create_config(body: dict):
    mgr = _get_mgr()
    config = await mgr.create_config(body)
    return success(config)


@router.get("/{configId}")
async def get_config(configId: str):
    mgr = _get_mgr()
    config = await mgr.get_config(configId)
    if not config:
        from app.core.exceptions import AppError, ERROR_NOT_FOUND
        raise AppError(ERROR_NOT_FOUND, f"配置 '{configId}' 不存在", status_code=404)
    return success(config)


@router.put("/{configId}")
async def update_config(configId: str, body: dict):
    mgr = _get_mgr()
    config = await mgr.update_config(configId, body)
    return success(config)


@router.delete("/{configId}")
async def delete_config(configId: str):
    mgr = _get_mgr()
    await mgr.delete_config(configId)
    return success()


@router.put("/{configId}/enabled")
async def toggle_enabled(configId: str, body: dict):
    mgr = _get_mgr()
    config = await mgr.set_enabled(configId, body.get("enabled", True))
    return success(config)


@router.post("/{configId}/test")
async def test_config(configId: str, body: dict):
    """Test execute an API config with given params."""
    mgr = _get_mgr()
    config = await mgr.get_config(configId)
    if not config:
        from app.core.exceptions import AppError, ERROR_NOT_FOUND
        raise AppError(ERROR_NOT_FOUND, f"配置 '{configId}' 不存在", status_code=404)

    # Delegate to query pipeline for execution
    pipeline = registry.get_optional("query_pipeline")
    if pipeline:
        import time
        start = time.time()
        result = await pipeline.execute(config, body.get("params", {}))
        duration = int((time.time() - start) * 1000)
        return success({**result, "executionTime": duration})

    return success({"message": "QueryPipeline 未就绪，请先启动 Mock Server"})

File: frontend/src/api/apiConfigs.ts
import client from './client';

// ==== API Configs ====

export const listConfigs = (params?: { keyword?: string; group?: string; enabled?: boolean; dataSourceType?: string }) =>
  client.get('/api-configs', { params });

export const getConfig = (id: string) =>
  client.get(`/api-configs/${id}`);

export const createConfig = (data: Record<string, unknown>) =>
  client.post('/api-configs', data);

export const updateConfig = (id: string, data: Record<string, unknown>) =>
  client.put(`/api-configs/${id}`, data);

export const deleteConfig = (id: string) =>
  client.delete(`/api-configs/${id}`);

export const toggleEnabled = (id: string, enabled: boolean) =>
  client.put(`/api-configs/${id}/enabled`, { enabled });

export const testConfig = (id: string, params: Record<string, unknown>) =>
  client.post(`/api-configs/${id}/test`, { params });

export const listGroups = () =>
  client.get('/api-configs/groups');

export const getDefaultSql = (collection: string) =>
  client.get(`/api-configs/default-sql/${collection}`);

export const exportConfigs = (group?: string) =>
  client.get('/api-configs/export', { params: group ? { group } : {} });

export const importConfigs = (configs: unknown[], overwrite?: boolean) =>
  client.post('/api-configs/import', { configs, overwrite });

// ==== Sessions ====

export const listSessions = () =>
  client.get('/sessions');

export const deleteSession = (token: string) =>
  client.delete(`/sessions/${token}`);

export const clearSessions = () =>
  client.delete('/sessions');

export const getSessionConfig = () =>
  client.get('/sessions/config');

export const updateSessionConfig = (data: Record<string, unknown>) =>
  client.put('/sessions/config', data);

// ==== Protocols ====

export const listProtocols = () =>
  client.get('/protocols');

export const startProtocol = (name: string) =>
  client.post(`/protocols/${name}/start`);

export const stopProtocol = (name: string) =>
  client.post(`/protocols/${name}/stop`);

export const restartProtocol = (name: string) =>
  client.post(`/protocols/${name}/restart`);

export const getProtocolHealth = (name: string) =>
  client.get(`/protocols/${name}/health`);

export const startAllProtocols = () =>
  client.post('/protocols/start-all');

export const stopAllProtocols = () =>
  client.post('/protocols/stop-all');

export const getMockConfig = () =>
  client.get('/protocols/http-mock/config');

export const updateMockConfig = (data: Record<string, unknown>) =>
  client.put('/protocols/http-mock/config', data);

export const listMockRoutes = () =>
  client.get('/protocols/http-mock/routes');

export const listRequestLogs = (params?: { limit?: number; offset?: number; method?: string; path?: string }) =>
  client.get('/protocols/http-mock/logs', { params });

export const clearRequestLogs = () =>
  client.delete('/protocols/http-mock/logs');

// ==== Settings ====

export const listSettings = (category?: string) =>
  client.get('/settings', { params: category ? { category } : {} });

export const updateSetting = (key: string, value: string) =>
  client.put(`/settings/${key}`, { value });

export const batchUpdateSettings = (settings: Array<{ key: string; value: string }>) =>
  client.put('/settings', { settings });
